### PR TITLE
Enable test duration writes on all runs

### PR DESCRIPTION
**Pull request by sondrelg**
This PR:

- Adds the possibility of outputting durations using `--store-durations` even when splitting. This seems useful for maintaining a "warm cache" of durations. If test fixtures change, or test environments are upgraded, test durations can now continuously update to reflect those changes.
- Add the possibility of running `pytest-split` with no durations file present. In that case we now just split tests evenly. This is a small detail, but one that will make implementing continuous updates in a ci-environment a little easier I think.
- Changes the saved data structure from a list of lists to a dict. This makes continuous updates easier since we can just update or add affected keys.

There's also a slight restructuring wrt. pytest plugin structure, which is not a big deal if you prefer to revert. Just let me know - I just thought it might make more sense to register cache/split functionality based on certain criteria rather than run it every time and return early when not relevant. At least that was what I was trying to do 🙂

### CODE REVIEW

**Comment by jerry-git** [comment]
Some thoughts based on the PR description (haven't looked at the code yet 🙂 ).

> Implements writes to the durations file on every run, even when splitting tests.

This is OK if it's still behind `--store-durations`. I.e. if that flag is not used, I would not store the durations.

> Moves the default test-duration storage file to the pytest cache at .pytest_cache/v/cache/pytest-split

This is problematic as basically all the projects have `.pytest_cache` in their gitignore.

**Comment by sondrelg** [comment]
> > Implements writes to the durations file on every run, even when splitting tests.
> 
> This is OK if it's still behind `--store-durations`. I.e. if that flag is not used, I would not store the durations.

Ok that sounds good 👍  I'll make that the default behavior

> > Moves the default test-duration storage file to the pytest cache at .pytest_cache/v/cache/pytest-split
>
> This is problematic as basically all the projects have .pytest_cache in their gitignore.

Yeah I should maybe have waited to let you respond in https://github.com/jerry-git/pytest-split/issues/11 before starting, but had some time yesterday so thought I'd jump into it 🙂  

Personally I would prefer that the cached durations be left *out* of version control, since the files generated can become quite big. At work we have a medium sized project with ~4000 tests which produces a 550kb sized test-durations output file, which is quite big and will become part of our PR reviews every time it changes. 

At the same time, there's no need to make this the default behavior if you don't want it to be - we could skip the pytest-cache stuff altogether, and I could just put the output file (`.test-durations`) in my .gitignore at work. Do you maybe want to reply to the posts in the issue first, and then we can progress from there?

**Comment by sondrelg** [comment]
I think I have a good idea of how to proceed now @jerry-git; thanks for responding in the issue. 

One final thing before finalizing this - would you mind quickly taking a look at the [_split_tests](https://github.com/jerry-git/pytest-split/blob/cd95f850b4eb175d3a415ec9a3cd49861dfadb26/src/pytest_split/plugin.py#L131) method I added? I had a little bit of trouble understanding the original split-function (looking at it now it seems makes sense, I think I was just tired), so I changed it slightly. 

Some of the changes are new and have to be there (I think), like:

- We need to set `avg_duration_per_test = 1` when no durations are found to support running tests when no durations file has been added yet. I guess this would be considered a new  backwards-compatible feature.

Then there's a change that's a consequence of changing how we select (and deselect) tests a little
```python
items[:] = selected_tests        
config.hook.pytest_deselected(items=deselected_tests)
```

The deselection results in pytest outputting something like `Found 45, selected 12, deselected 33` which I think is a slight improvement 🙂 

The final change is that instead of finding the index at which the runtime threshold is reached, I've now sorted the tests by slowest first and then add faster tests to get as close to the threshold as possible. It might be overkill and could be reverted if you don't see any value in it.

If you could review that function specifically I think I have everything I need to finish 👍

**Comment by jerry-git** [comment]
> We need to set avg_duration_per_test = 1 when no durations ...

Sounds legit 👍 

> The deselection results in pytest outputting something like Found 45, selected 12, deselected 33 which I think is a slight improvement

👍 

> I've now sorted the tests by slowest first and then add faster tests to get as close to the threshold as possible

One of the key features of `pytest-split` is that it maintains the order of tests, just make sure we don't lose that 🙂  One future feature idea would be to randomise based on some flag (i.e. something similar as [pytest-randomly](https://github.com/pytest-dev/pytest-randomly) is doing) but let's maintain the alphabetical order by default.

**Comment by sondrelg** [comment]
Sounds good 👍  🙂

**Comment by sondrelg** [comment]
I'm more or less done I think 🙂  Just have one quick question @jerry-git: what do you think about making the pytest-split output stand out a little?

This is all the output messages I've left in currently (the top one will only be output when no durations file is found)


image

It might not be a big deal, but right now it might be hard to differentiate between regular pytest output and plugin output.

How do you feel about doing changing the color of the output slightly? Like this (bold white):

image

Or this (yellow)? 😄 

image


The last one might be a little much - a color change might not be necessary - perhaps adding a `>> ` to the start of the output strings or an equivalent marker would have the same effect.

Might also not be necessary at all, just thought I'd ask 🙂

File: tests/test_plugin.py
Comment (sondrelg) [inline tests/test_plugin.py:28]: I had to change all the expected data in the tests to be dicts instead of lists

**Comment by sondrelg** [comment]
A lot of back and forth here, but I think the PR is now ready for a second look when you have time @jerry-git 🙂

**Comment by mbkroese** [comment]
@sondrelg this PR will have a lot of conflicts with this one: https://github.com/jerry-git/pytest-split/pull/12

In that PR I'm basically adding the extra printouts that you're also adding in this PR. I also change the logic somewhat to use a different PyTest hook. I agree the idea for the colours is nice, but I think it would be better to keep these changes in separate PRs. What do you think about commenting on my PR, so we can merge that, and afterwards apply your PR on top of it?

**Comment by sondrelg** [comment]
@mbkroese that sounds good to me 👍 

Sorry I didn't see it before - not my intention to create conflicts 🙂 If you and Jerry can agree on how to resolve #12 I'll pull master and resolve conflicts afterwards.

File: src/pytest_split/plugin.py
Comment (jerry-git) [inline src/pytest_split/plugin.py:18]: Is this still relevant?

**Comment by jerry-git** [comment]
What happens in a situation in which I'd have the `.test_durations` file in the old format and I'd run with a version of pytest-split which has this changeset present?

File: src/pytest_split/plugin.py
Comment (jerry-git) [inline src/pytest_split/plugin.py:77]: I like the change to two separate plugins and usage of `config.pluginmanager` 👍

**Comment by sondrelg** [comment]
> What happens in a situation in which I'd have the `.test_durations` file in the old format and I'd run with a version of pytest-split which has this changeset present?

The implementation would fail 🙂 

Adding handling for this would be easy though - should we just re-format the data on-read if we see the old format and write with the new format?

**Comment by sondrelg** [comment]
I have an unrelated question for you btw @jerry-git - not directly related to this PR, just something I was curious about and I suspect you're the best person to ask 🙂  

Do you think one should expect test-times to vary considerably for individual tests between runs? Or is it reasonable to expect that a single test should run with the same absolute duration, or relative duration to other tests? 

I tried testing this PR on another project. So for each run I'm writing durations - that works well; but the splits are still pretty uneven:

image

image

image

The ideal outcome for me would be that every block runs for ~9 minutes, so getting some blocks with 6 minute runs and some with 12 minute runs isn't ideal 😛  Not sure whether this happens because of inconsistencies in Github Actions, or if there is something we could do to the split logic to improve things. Have you had similar experiences?

I've not gotten around to it, but I will eventually collect the durations for a few different runs and actually just look at the data - I'm sure that will shed some light on things - so mostly just asking in case you've already thought about/looked into this in the past.

**Comment by mbkroese** [comment]
@sondrelg I'm using this plugin on a test suite of ~1500 tests and noticed a similar problem.

# the splitting algorithm
The current splitting algorithm can easily be fooled into creating uneven groups. I'm planning to submit a PR shortly to update this to a more optimal algorithm. I'll create a PR for that once the current open PRs are merged.

# setup time
Some of my tests have large setup times. I _think_ those setup times are included in the test durations. This gives the impressions that some tests are slow when in fact they are not.

# variability
Our tests also have variability in testing times (for various reasons). Perhaps we could account for this by storing test durations and taking an average/median/min/max?

**Comment by sondrelg** [comment]
Cool, that's more or less exactly what I was thinking 👏  

If the setup time is recorded as runtimes I'm guessing that would immediately fix 90%+ of my issue. Our database migrations take about 2-3 minutes to run, so that would be a perfect explanation as to why our splits are uneven. I think I'll take a look at the durations file later today if I get time - it should be easy enough to verify 👍  I actually hope this is the case, as this would be the easiest possible explanation to fix! 🤞 

The reason I asked was actually to suggest recording an array of durations instead of a single duration as a possible improvement, to account for variability. This also seems like a very simple thing to implement - thought it might be worthwhile to limit the length of the arrays to a low-ish max-length and rotate durations out, so the durations file doesn't keep growing.

Exciting 👍  

I won't do anything new with this PR though, I'll just wait until #12 is merged, fix conflicts, and fix the backwards compatibility issue + any new feedback you might have 🙂

**Comment by sondrelg** [comment]
Looked up the durations and yeah it looks like the start-time is recorded:

These are the tests over 100s in my durations file with the recorded time and the time it took to re-run them locally.

```
::test 243.08475018500008 - 2.63s
::test 233.48524261500006 - 2.14s
::test[variation1] 202.96748632899997 -
::test[variation2] 202.61207506399998 - 8.51s
::test 237.58321302000002 - 4.41
::test 236.68534560499998 - 4.53
::test 190.07624340700005 - 5.36
::test 190.2456836280001 - 5.34
::test 194.08574962100002 - 4.46
::test 194.78907412200005 - 4.31
```

For some reason I have 10 (9) tests with long runtimes, when I split in 8 groups 🤔  But looks like a good explanation of the unevenness!

**Comment by jerry-git** [comment]
>> What happens in a situation in which I'd have the .test_durations file in the old format and I'd run with a version of pytest-split which has this changeset present?

> The implementation would fail 🙂
Adding handling for this would be easy though - should we just re-format the data on-read if we see the old format and write with the new format?

I see two options:
1. We accept the non-backwards-compatibility and communicate it clearly (e.g. with some release notes).
2. Reading would support both the old and the new format. Writing would always use the new format. However, I would only write when `--store-durations` is used.

1. would be easiest and 2. would be most user-friendly. If 2 would be used, there could be some EOL for the old format support so that the code could be simplified at some point.

Any thoughts?

**Comment by jerry-git** [comment]
Great insights with the durations! I assume @sondrelg's findings are related to running db migrations in some fixture once per sub suite?

**Comment by jerry-git** [comment]
> I won't do anything new with this PR though, I'll just wait until #12 is merged, fix conflicts, and fix the backwards compatibility issue + any new feedback you might have 🙂

Great! I'll merge #12 so we can move forward 🙂  Awesome work and knowledge sharing from both @sondrelg & @mbkroese! 💎

**Comment by sondrelg** [comment]
> I see two options:
> 
> 1. We accept the non-backwards-compatibility and communicate it clearly (e.g. with some release notes).
> 2. Reading would support both the old and the new format. Writing would always use the new format. However, I would only write when `--store-durations` is used.
> 3. would be easiest and 2. would be most user-friendly. If 2 would be used, there could be some EOL for the old format support so that the code could be simplified at some point.
> 
> Any thoughts?

I don't mind adding support for reading both. I can add a comment to remember to remove support for the old format when you release v1 for example 🙂  

> Great insights with the durations! I assume @sondrelg's findings are related to running db migrations in some fixture once per sub suite?

Yep! [This](https://github.com/pytest-dev/pytest-django/blob/master/pytest_django/fixtures.py#L96) fixture is what takes so long, and it looks like the way pytest is set up, the setup times of all fixtures (including those scoped for a whole session) are attributed to the first test in the test suite. I opened [this](https://github.com/pytest-dev/pytest/issues/8741) issue in the pytest repo; hopefully it's possible to fix this there in time. 

In the meantime though, I would suggest pytest-split should do something like [this](https://github.com/sondrelg/pytest-split/blob/start-up-fix/src/pytest_split/plugin.py#L195) where we just remove the setup time of the first test in the test suite. 

I'm testing that branch on my project at work now and it seems to have gotten rid of a lot of the variability 🎉 

If you're keen to work on the split logic and variability fixes after this PR is merged, feel free to tag me in the PR and I'd be happy to test/review it for you when it's ready 🙂

**Comment by jerry-git** [comment]
> In the meantime though, I would suggest pytest-split should do something like this where we just remove the setup time of the first test in the test suite.

This sounds like a potential solution but how about if the first test doesn't use the heavy fixture? 🙂

**Comment by sondrelg** [comment]
PR should be ready to review when you're ready @jerry-git 👏

**Comment by jerry-git** [comment]
> but we know that all session scoped fixtures will always be attributed to the first test

This is not true, consider e.g.
```python
import time
import pytest

@pytest.fixture(scope="session")
def expensive_session_fixture():
    time.sleep(10)

def test_1():
    ...

def test_2(expensive_session_fixture):
    ...
```

Yup, we could have some flag for enabling / disabling the "skip measuring the first setup". Considering your use case (and the whole thing in general), I think more valuable flag could be something like `--discard-long-durations 20` which would basically mean that it doesn't store durations for tests which are measured to last more than 20 seconds.

**Comment by sondrelg** [comment]
Would you only apply the long-duration threshold for the `setup` time on the first test still?

**Comment by jerry-git** [comment]
> Would you only apply the long-duration threshold for the setup time on the first test still?

I would treat every test identically 🙂 

A configurable threshold would also provide a workaround against misc bugs with other libraries, see e.g. https://github.com/jerry-git/pytest-split/issues/1

File: src/pytest_split/plugin.py
Comment (jerry-git) [inline src/pytest_split/plugin.py:114]: 💪 

Is there a test case for this?

Comment (sondrelg) [reply src/pytest_split/plugin.py:114]: I'll add one 👍

Comment (jerry-git) [reply src/pytest_split/plugin.py:114]: Should probably add a 100% coverage requirement as well at some point 🤔

Comment (sondrelg) [reply src/pytest_split/plugin.py:114]: Do it tomorrow 😂 

Added the old format to test in 794fedf

Comment (sondrelg) [reply src/pytest_split/plugin.py:114]: I could help you write some tests if you want, and if you'd like I could also help implement [poetry](https://python-poetry.org/) for package management and dependency stuff 🙂

Comment (jerry-git) [reply src/pytest_split/plugin.py:114]: Poetry would be welcome indeed 👍

**Review by jerry-git — APPROVED** [review]
💪

**Comment by jerry-git** [comment]
FYI, available in [0.2.0](https://github.com/jerry-git/pytest-split/releases/tag/0.2.0)!

**Comment by mbkroese** [comment]
> collected 16 items / 7 deselected / 9 selected 

This looks really good 👍

**Pull request merged**
